ZIPs 32 and 316: Refine how IVK components are derived, and other cleanups (#564)
zcash/zips: zip-0316.rst

Receivers, and therefore the Priority List defined in the previous
section is not used.

For each FVK Type or IVK Type currently defined in this specification,
the same Typecode is used as for the corresponding Receiver Type in a
Unified Address. Additional FVK Types and IVK Types MAY be defined in
future, and these will not necessarily use the same Typecode as the
corresponding Unified Address.

The following FVK or IVK Encodings are used in place of the
:math:`\mathtt{addr}` field:

* An Orchard FVK or IVK Encoding, with Typecode :math:`\mathtt{0x03},` is
  is the raw encoding of the Orchard Full Viewing Key or Orchard Incoming
  Viewing Key respectively.

* A Sapling FVK Encoding, with Typecode :math:`\mathtt{0x02},` is
  the encoding of a Sapling Extended Full Viewing Key defined in
  [#zip-0032-sapling-extfvk]_. This SHOULD be an Extended Full Viewing

>   Key at the Account level of the ZIP 32 hierarchy.

Reviewer: In ZIP sync today, we agreed to change this to instead encode Sapling FVKs using the `EncodeExtFVKParts` function from ZIP 32, so they are a representation of `(ak, nk, ovk, dk)`. This reduces the size of the encoding by 25%, and removes capabilities (`c`) and potentially-private information (child index which is the account number) from the FVK encoding.

The downside is that we can't extract a Sapling ExtFVK from a UFVK losslessly, but that doesn't matter in most cases. If someone needs to do that, they should set `(depth, parent_fvk_tag, i, c)` to the string of all 1 bits, to indicate "data unknown".

  treated as unrecognized by Consumers.

* For Transparent P2PKH Addresses that are derived according to BIP 32
  [#bip-0032]_ and BIP 44 [#bip-0044]_, the FVK and IVK Encodings
  have Typecode :math:`\mathtt{0x00}.` Both of these encodings are the
  serialization of an extended public key, defined in the section

>   “Serialization format” of BIP 32 [#bip-0032-serialization-format]_.

Reviewer: In ZIP sync today, we agreed to alter this to instead just encode:
- The chain code.
- The public key data.

This is effectively the suffix of the extended public key encoding, leaving out the potentially-private data (e.g. child index which is the account). The chain code `c` is necessary to include, as we leverage the capability to derive children as a transparent diversification mechanism.

  treated as unrecognized by Consumers.

* For Transparent P2PKH Addresses that are derived according to BIP 32
  [#bip-0032]_ and BIP 44 [#bip-0044]_, the FVK and IVK Encodings have
  Typecode :math:`\mathtt{0x00}.` Both of these are encodings of the
  chain code and public key :math:`(\mathsf{c}, \mathsf{pk})` given by
  :math:`\mathsf{c}\,||\,\mathsf{ser_P}(\mathsf{pk})`. (This is the
  same as the last 65 bytes of the extended public key format defined
  in section “Serialization format” of BIP 32 [#bip-0032-serialization-format]_.)
  However, the FVK uses the key at the Account level, i.e. at path

>   :math:`m / 44' / coin\_type' / account'`, while the IVK uses its
>   non-change child key at path :math:`m / 44' / coin\_type' / account' / 0`.

Reviewer 2: ```suggestion
  :math:`m / 44' / coin\_type' / account'`, while the IVK uses the
  external child key at the change path level 
  :math:`m / 44' / coin\_type' / account' / 0`.
```

  all less than 256 bytes and so could use a one-byte length field,
  encodings for experimental types may be longer.)

> * Each Receiver and FVK SHOULD represent an Address or Full Viewing
>   Key at the ZIP 32 or BIP 44 Account level.
> 
> * Each IVK SHOULD represent an Incoming Viewing Key at the Account
>   level for Sapling or Orchard, or its non-change child extended public
>   key for Transparent P2PKH.

Reviewer: I don't think this change makes sense to include, because it's a) no longer requirements that apply to both UAs and UVKs, and b) I think both the original requirement and this change are just duplicating the earlier definitions of FVK and IVK encodings.

However, I think what we _should_ have is a requirement of the following form, which I don't believe is currently stated anywhere else in ZIP 316:
```suggestion
* Within a single UA or UVK, any HD-derived Receiver, FVK, or IVK SHOULD
  represent an Address or Viewing Key for the same account (as used in
  the ZIP 32 or BIP 44 Account level).
```

Reviewer 2: Maybe instead of "any HD-derived Receiver, FVK, or IVK" it should read "all HD-derived Receivers, FVKs, and IVKs"

short-term, and should otherwise be either standardized in a ZIP (and
allocated a Typecode outside this reserved range) or discontinued.

New types SHOULD maintain the same distinction between FVK and IVK
authority as existing types, i.e. an FVK is intended to give access to
view all transactions to and from the address, while an IVK is intended
to give access only to view incoming payments (as opposed to change).

Deriving a UIVK from a UFVK
---------------------------

The following derivations are applied to each component FVK:

* For a Sapling FVK, the corresponding Sapling IVK is obtained as
  specified in [#protocol-saplingkeycomponents]_.

* For an Orchard FVK, the corresponding Orchard IVK is obtained as
  specified in [#protocol-orchardkeycomponents]_.

* For a Transparent P2PKH FVK, the corresponding Transparent P2PKH IVK
  is obtained from the extended public key corresponding to the account's
  non-change key path, i.e. by deriving the child key with non-hardened

>   index :math:`0`. It is encoded in the same way as the FVK Encoding.

Reviewer 2: This is no longer quite correct, since the FVK doesn't contain the extended public key; this should probably refer instead to the function `CKDpub((Kpar, cpar), i) → (Ki, ci)` as defined in BIP 32.

  defined in ZIP 32 section “Sapling diversifier derivation”
  [#zip-0032-sapling-diversifier-derivation]_.
* A Transparent diversifier index MUST be in the range :math:`0` to
  :math:`2^{31} - 1` inclusive.

There are no additional constraints on an Orchard diversifier index.

In the case of deriving a Transparent P2PKH Receiver from a Transparent
P2PKH IVK, the diversifier index is used as a BIP 44 child key index
at the Index level [#bip-0044-path-index]_ to derive the address. As is
usual for derivations below the BIP 44 Account level, non-hardened

> (public) derivation [#bip-0032-public-to-public]_ MUST be used. The IVK

Reviewer 2: This also uses the outdated definition of the IVK.

> is assumed to be the extended public key for the non-change element of
> the path. That is, if the UIVK was constructed correctly then the BIP 44
> path of the Transparent P2PKH Receiver will be
> :math:`m / 44' / \mathit{coin\_type\kern0.05em'} / \mathit{account\kern0.1em'} / 0 / \mathit{diversifier\_index}.`

Reviewer 3: ```suggestion
is assumed to correspond to the extended public key for the non-change
element of the path. That is, if the UIVK was constructed correctly then
the BIP 44 path of the Transparent P2PKH Receiver will be
:math:`m / 44' / \mathit{coin\_type\kern0.05em'} / \mathit{account\kern0.1em'} / 0 / \mathit{diversifier\_index}.`
```
